extensions/amp-user-location/0.1/config-loader.js

```diff
@@ -0,0 +1,199 @@
+    }
+
+    if (this.hasScriptConfig_() && this.hasRemoteConfig_()) {
+      // Prefer the fetchConfig if it's ready, and if not use the script config.
+      return Promise.race([this.fetchConfig(), this.getScriptConfig_()]);
+    }
+
+    // this.hasScriptConfig_() must have returned false by this point
+    if (this.hasRemoteConfig_()) {
+      return this.fetchConfig();
+    }
+
+    // If there is no remote or local config, use an empty dict.
+    return Promise.resolve(dict());
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  hasScriptConfig_() {
+    return this.element_.children.length > 0;
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  hasRemoteConfig_() {
+    return this.element_.hasAttribute('src');
+  }
+
+  /**
+   * Parse a JSON script tag for configuration, if present.
+   * @return {!Promise<!JsonObject>}
+   * @private
+   */
+  getScriptConfig_() {
+    if (!this.hasScriptConfig_()) {
+      return Promise.resolve(dict());
+    }
+
+    if (this.scriptConfigPromise_) {
+      return this.scriptConfigPromise_;
+    }
+
+    const {children} = this.element_;
+    userAssert(
+      children.length == 1,
+      'amp-user-location may only have one configuration json <script> tag'
+    );
+
+    const firstChild = children[0];
+    if (!isJsonScriptTag(firstChild)) {
+      user().error(
+        TAG,
+        'amp-user-location config should be in a <script> tag ' +
+          'with type="application/json".'
+      );
+      return Promise.resolve(dict()); // TODO(cvializ): error handling is hard : (
+    }
+
+    const {promise, resolve, reject} = new Deferred();
+    const json = tryParseJson(firstChild.textContent, e => {
+      // Rejects synchronously, so if an error occurs,
+      // the resolve below will have no effect
+      reject(e);
+    });
+    resolve(assertIsObject(json));
+
+    return (this.scriptConfigPromise_ = promise);
+  }
+
+  /**
+   * If a remote config is present, fetch it. Consumer code can call this
+   * early to pre-fetch the config before calling `getConfig`.
+   * This is designed to only fetch the config once.
+   * @return {!Promise<!JsonObject>}
+   */
+  fetchConfig() {
+    if (!this.hasRemoteConfig_()) {
+      return Promise.resolve(dict());
+    }
+
+    if (this.fetchPromise_) {
+      return this.fetchPromise_;
+    }
+
+    return (this.fetchPromise_ = this.getRemoteConfig_());
+  }
+
+  /**
+   * Reads the data from the URL provided in the 'src' attribute.
+   * For use with remote data.
+   * @return {!Promise<!JsonObject>}
+   * @private
+   */
+  getRemoteConfig_() {
+    const policy = UrlReplacementPolicy.ALL;
+    return this.fetch_(this.ampdoc_, this.element_, '.', policy);
+  }
+
+  /**
+   * Wrapper to stub during testing.
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!Element} element
+   * @param {string} token
+   * @param {!UrlReplacementPolicy} policy
+   * @return {!Promise<!JsonObject>}
+   * @private
```

Comment:
Would do `@package Visible for testing`. We currently do not run type check against our tests, but should still type things as if we were.